block-storage/src/main/scala/coop/rchain/blockstorage/LMDBBlockStore.scala

```diff
@@ -0,0 +1,128 @@
+                blocks.put(txn,
+                           blockHash.toDirectByteBuffer,
+                           blockMessage.toByteString.toDirectByteBuffer)
+                txn.commit()
+              }
+            }(txn => syncF.delay(txn.close()))
+    } yield ret
+
+  def get(blockHash: BlockHash): F[Option[BlockMessage]] =
+    for {
+      _ <- metricsF.incrementCounter(MetricNamePrefix + "get")
+      ret <- syncF.bracket(syncF.delay(env.txnRead()))(txn =>
+              syncF.delay {
+                val r = Option(blocks.get(txn, blockHash.toDirectByteBuffer)).map(r =>
+                  BlockMessage.parseFrom(ByteString.copyFrom(r).newCodedInput()))
+                txn.commit()
+                r
+            })(txn => syncF.delay(txn.close()))
+    } yield ret
+
+  def asMap(): F[Map[BlockHash, BlockMessage]] =
+    for {
+      _ <- metricsF.incrementCounter(MetricNamePrefix + "as-map")
+      ret <- syncF.bracket(syncF.delay(env.txnRead()))(txn =>
+              syncF.delay {
+                val r = blocks.iterate(txn).asScala.foldLeft(Map.empty[BlockHash, BlockMessage]) {
+                  (acc: Map[BlockHash, BlockMessage], x: CursorIterator.KeyVal[ByteBuffer]) =>
+                    val hash = ByteString.copyFrom(x.key())
+                    val msg  = BlockMessage.parseFrom(ByteString.copyFrom(x.`val`()).newCodedInput())
+                    acc.updated(hash, msg)
+                }
+                txn.commit()
+                r
+            })(txn => syncF.delay(txn.close()))
+    } yield ret
+
+}
+
+object LMDBBlockStore {
+
+  private val MetricNamePrefix = "lmdb-block-store-"
+
+  def create[F[_]](env: Env[ByteBuffer], path: Path)(implicit
+                                                     syncF: Sync[F],
+                                                     metricsF: Metrics[F]): BlockStore[F] = {
+
+    val blocks: Dbi[ByteBuffer] = env.openDbi(s"blocks", MDB_CREATE)
+    new LMDBBlockStore(env, path, blocks)
+  }
+
+  def createWithId(env: Env[ByteBuffer], path: Path): BlockStore[Id] = {
+    import coop.rchain.metrics.Metrics.MetricsNOP
+    val sync: Sync[Id] =
+      new Sync[Id] {
+        def pure[A](x: A): cats.Id[A] = x
+
+        def handleErrorWith[A](fa: cats.Id[A])(f: Throwable => cats.Id[A]): cats.Id[A] =
+          try { fa } catch {
+            case e: Exception => f(e)
+          }
+
+        def raiseError[A](e: Throwable): cats.Id[A] = throw e
+
+        def flatMap[A, B](fa: cats.Id[A])(f: A => cats.Id[B]): cats.Id[B] =
+          implicitly[FlatMap[Id]].flatMap(fa)(f)
+        def tailRecM[A, B](a: A)(f: A => cats.Id[Either[A, B]]): cats.Id[B] =
+          implicitly[FlatMap[Id]].tailRecM(a)(f)
+
+        def bracketCase[A, B](acquire: A)(use: A => B)(
+            release: (A, ExitCase[Throwable]) => Unit): B = {
+          val state = acquire
+          try {
+            use(state)
+          } finally {
+            release(acquire, ExitCase.Completed)
```

Comment:
I _think_ you could use different `ExitCase`s to release resources differently. In case of an error thrown you can `catch` it and release with `ExitCase.Error` which in turn could roll back the block store.